### PR TITLE
Remove table_lock during a training run

### PR DESCRIPTION
**Pull request by phillipleblanc**
Part of #392 

To improve the performance (speed) of training, it was important to remove acquiring the lock `table_lock` on every step of the training run. This lock was used to control when data was added to the data_manager for the pod to prevent conflicts where the table is being updated at the same time we are querying it during training.

The solution proposed here will remove the need for the `table_lock` during training by copying the data to be used during a training run. Data that is added to the data_manager will continue to be made available immediately for inferencing.

A secondary change that speeds up the processing of the `AddData` request is to not take a lock when adding data. Incoming data will be resampled to match the indexes of the existing table. Merging of the table has been improved to use `pandas.concat` followed by a `resample` which will not touch existing indexes.

### CODE REVIEW

File: ai/src/data_manager/base_manager.py [hunk 1]
```diff
@@ -64,6 +64,32 @@ def flatten_and_normalize_window(current_window) -> np.ndarray:
...
+    def __enter__(self):
+        """
+        Python Docs: https://docs.python.org/3/library/stdtypes.html#contextmanager.__enter__
+
+        Used to indicate to the datamanager that the pod is currently training. This should
+        trigger the data manager to make a deep copy of the underlying data to be used for training.
+        """
+        self.start_training()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
```
Comment (Corentin-pro) [inline ai/src/data_manager/base_manager.py]: pylint should complains with unused variable, we can prefix with `_` like : `def __exit__(self, _exc_type, _exc_val, _exc_tb):`

**Review by Corentin-pro — CHANGES_REQUESTED** [review]

**Review by Corentin-pro — APPROVED** [review]

**Comment by lukekim** [comment]
Nice! Did we get any before and after performance measurements?

File: test/e2e/e2e_test.go [hunk 2]
```diff
@@ -842,6 +842,8 @@ func TestImportExport(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	time.Sleep(time.Second * 5)
```
Comment (lukekim) [inline test/e2e/e2e_test.go]: Would be great to figure out why we need this.

**Review by lukekim — APPROVED** [review]

File: test/e2e/e2e_test.go (hunk 2 above)
Comment (phillipleblanc) [reply test/e2e/e2e_test.go]: The `ImportExport` test is now flaky and my theory was that adding a delay between when we start the runtime and import the pod to when we get an inference would help - it, unfortunately, does not. so I'll remove this delay and keep debugging.

**Comment by phillipleblanc** [comment]
EDIT: The following results are not conclusive - the spice and spiced binaries were from v0.4 but because I was using the metal context it was still using the version of the AI Engine with the lock improvements. I am re-running the tests

I did a test run of a spicepod that was taking a few minutes to train a single episode:

## v0.4-alpha (current release)
Episode Start: 14:37:37
Episode End: 14:42:33

Time taken: 4m56s -> 296 seconds
Screen Shot 2021-11-22 at 2 47 53 PM

## This change
Episode Start: 14:30:28
Episode End: 14:35:13

Time taken: 4m45s -> 285 seconds

Screen Shot 2021-11-22 at 2 50 28 PM

Speedup = 1 - (285/296) = 0.037 = 3.7%

**Comment by phillipleblanc** [comment]
Test in Docker:

Using a spicepod where a single episode trains in about 1 minute, run 5 episodes, and take the median time to complete.

## v0.4-alpha (current release)
5 episodes
Training start: 16:16:58
1: 16:16:58 - 16:18:03 = 64 seconds
2: 16:18:03 - 16:19:07 = 64 seconds
3: 16:19:07 - 16:20:12 = 65 seconds
4: 16:20:12 - 16:21:16 = 64 seconds
5: 16:21:16 - 16:22:20 = 64 seconds
Training end: 16:22:20

Median: 64 seconds

## PR 451
5 episodes
Training start: 16:06:40
1: 16:06:40 - 16:07:37 = 57 seconds
2: 16:07:37 - 16:08:36 = 59 seconds
3: 16:08:36 - 16:09:32 = 56 seconds
4: 16:09:32 - 16:10:26 = 54 seconds
5: 16:10:26 - 16:11:19 = 53 seconds
Training end: 16:11:19

Median: 56 seconds

12.5% speedup

**Comment by phillipleblanc** [comment]
# Memory stats

## v0.4-alpha
Start Training: RSS: 278.4296875mb

Start Training: GC Stats: [{'collections': 454, 'collected': 1813, 'uncollectable': 0}, {'collections': 41, 'collected': 117, 'uncollectable': 0}, {'collections': 5, 'collected': 0, 'uncollectable': 0}]

Start Training: GC Count: (15, 0, 0)

1: Episode End: RSS: 299.11328125mb
2: Episode End: RSS: 303.046875mb
3: Episode End: RSS: 307.5703125mb
4: Episode End: RSS: 311.89453125mb
5: Episode End: RSS: 316.25mb

End Training: RSS: 324.6875mb
End Training: GC Stats: [{'collections': 545, 'collected': 1813, 'uncollectable': 0}, {'collections': 49, 'collected': 117, 'uncollectable': 0}, {'collections': 7, 'collected': 4465, 'uncollectable': 0}]

End Training: GC Count: (24, 0, 0)

## New change
Start Training: RSS: 277.26171875mb

Start Training: GC Stats: [{'collections': 454, 'collected': 1813, 'uncollectable': 0}, {'collections': 41, 'collected': 117, 'uncollectable': 0}, {'collections': 5, 'collected': 0, 'uncollectable': 0}]

Start Training: GC Count: (12, 0, 0)

1: Episode End: RSS: 297.87109375mb
2: Episode End: RSS: 302.0078125mb
3: Episode End: RSS: 306.5625mb
4: Episode End: RSS: 310.98828125mb
5: Episode End: RSS: 315.4296875mb

End Training: RSS: 324.78515625mb
End Training: GC Stats: [{'collections': 546, 'collected': 1813, 'uncollectable': 0}, {'collections': 49, 'collected': 117, 'uncollectable': 0}, {'collections': 7, 'collected': 4465, 'uncollectable': 0}]

End Training: GC Count: (25, 0, 0)

**Pull request merged**